pkg/planner/core/operator/logicalop/logical_aggregation.go

```diff
@@ -632,44 +632,96 @@ func (la *LogicalAggregation) splitCondForAggregation(predicates []expression.Ex
+	for idx, col := range la.Schema().Columns {
+		aggFunc := la.AggFuncs[idx]
+		if aggFunc.Name == ast.AggFuncFirstRow {
+			cols := expression.ExtractColumns(aggFunc.Args[0])
+			if len(cols) == 1 {
+				aggFuncsCols = append(aggFuncsCols, col)
+			}
+		}
+	}
+	return aggFuncsCols
+}
+
+// pushDownPredicates split a condition to two parts, can be pushed-down or can not be pushed-down below aggregation.
+func (*LogicalAggregation) pushDownPredicates(cond expression.Expression, groupByColumns *expression.Schema, exprsOriginal []expression.Expression, pushDownFunc pushDownPredicatesFunc) (condsToPush, ret []expression.Expression) {
+	switch c := cond.(type) {
 	case *expression.Constant:
 		condsToPush = append(condsToPush, cond)
 		// Consider SQL list "select sum(b) from t group by a having 1=0". "1=0" is a constant predicate which should be
 		// retained and pushed down at the same time. Because we will get a wrong query result that contains one column
 		// with value 0 rather than an empty query result.
 		ret = append(ret, cond)
 	case *expression.ScalarFunction:
-		extractedCols := expression.ExtractColumns(cond)
-		ok := true
-		for _, col := range extractedCols {
-			if !groupByColumns.Contains(col) {
-				ok = false
-				break
-			}
+		return pushDownFunc(c, groupByColumns, exprsOriginal)
+	default:
+		ret = append(ret, cond)
+	}
+	return condsToPush, ret
+}
+
+type pushDownPredicatesFunc func(cond *expression.ScalarFunction, cols *expression.Schema, exprsOriginal []expression.Expression) (_, _ []expression.Expression)
+
+func (la *LogicalAggregation) pushDownPredicatesByGroupby(cond *expression.ScalarFunction, groupByColumns *expression.Schema, exprsOriginal []expression.Expression) (condsToPush, ret []expression.Expression) {
+	extractedCols := expression.ExtractColumns(cond)
+	ok := true
+	for _, col := range extractedCols {
+		if !groupByColumns.Contains(col) {
+			ok = false
+			break
 		}
-		if ok {
-			newFunc := expression.ColumnSubstitute(la.SCtx().GetExprCtx(), cond, la.Schema(), exprsOriginal)
-			condsToPush = append(condsToPush, newFunc)
-		} else {
-			ret = append(ret, cond)
+	}
+	if ok {
+		newFunc := expression.ColumnSubstitute(la.SCtx().GetExprCtx(), cond, la.Schema(), exprsOriginal)
+		condsToPush = append(condsToPush, newFunc)
+	} else {
+		ret = append(ret, cond)
+	}
+	return
+}
+
+// pushDownPredicatesByAggFuncs is only used for firstrow
+func (la *LogicalAggregation) pushDownPredicatesByAggFuncs(cond *expression.ScalarFunction, aggFirstRowColumns *expression.Schema, exprsOriginal []expression.Expression) (condsToPush, ret []expression.Expression) {
+	extractedCols := expression.ExtractColumnsMapFromExpressions(nil, cond)
+	var schemaCol *expression.Column
+	if len(extractedCols) != 1 {
+		ret = append(ret, cond)
+		return condsToPush, ret
+	}
+	for _, col := range extractedCols {
+		schemaCol = aggFirstRowColumns.RetrieveColumn(col)
+		if schemaCol == nil {
```

Comment:
Should be `if schemaCol != nil {break}`?

Reply:
Yes， it have been refactored.